### PR TITLE
Add software tonemap filter support

### PR DESCRIPTION
**Pull request by gnattu**
This adds a fast SIMD optimized software tonemap filter support to the server. This filter currently does not support dolby vision reshaping and only support HDR10 and HLG inputs. With current veryfast libx264 settings we can expect 4K60 transcoding+tone mapping on a modern processor with 8 cores equivalent to Zen3/Golden Cove/Firestorm level performance. 

**Changes**

- Add tonemapx filter support to software transcode pipeline
- Add tonemapx filter support to thumbnail generation pipeline

**Issues**

Depends on ffmpeg pr https://github.com/jellyfin/jellyfin-ffmpeg/pull/407

The perf figures in the original post only reflects the initial implementation. The finalized perf figures are here: https://github.com/jellyfin/jellyfin-ffmpeg/pull/407#issuecomment-2198303178

### CODE REVIEW

**Review by Bond-009 — APPROVED** [review]

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 1]
```diff
@@ -276,6 +276,21 @@ private bool IsVideoToolboxFullSupported()
                 && _mediaEncoder.SupportsFilter("scale_vt");
         }
 
+        private bool IsSwTonemapAvailable(EncodingJobInfo state, EncodingOptions options)
+        {
+            if (state.VideoStream is null
+                || !options.EnableTonemapping
+                || GetVideoColorBitDepth(state) != 10
+                || !_mediaEncoder.SupportsFilter("tonemapx")
+                || !(string.Equals(state.VideoStream?.ColorTransfer, "smpte2084", StringComparison.OrdinalIgnoreCase) || string.Equals(state.VideoStream?.ColorTransfer, "arib-std-b67", StringComparison.OrdinalIgnoreCase)))
```
Comment (nyanmisaka) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:285]: Are these checks duplicated?
```
|| !(string.Equals(state.VideoStream?.ColorTransfer, "smpte2084", StringComparison.OrdinalIgnoreCase) || string.Equals(state.VideoStream?.ColorTransfer, "arib-std-b67", StringComparison.OrdinalIgnoreCase)))
```
```
            return state.VideoStream.VideoRange == VideoRange.HDR
                   && state.VideoStream.VideoRangeType is VideoRangeType.HDR10 or VideoRangeType.HLG or VideoRangeType.DOVIWithHDR10 or VideoRangeType.DOVIWithHLG;
```

**Review by nyanmisaka — APPROVED** [review]
Speaking of Golden Cove, let's hope that the flagship models in Raptor Lake don't crash in our AVX2 code.

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs (hunk 1 above)
Comment (gnattu) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:285]: The former checks for all supported HDR curves and the latter filters out Dolby vision videos without fallback profiles. Only checks for the latter should be enough but we are doing this double check for hardware tone mapping as well.

Comment (nyanmisaka) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:285]: https://github.com/jellyfin/jellyfin/blob/eb45a36af0ff3ef540a4a936ea1329139f0523c1/MediaBrowser.Model/Entities/MediaStream.cs#L703-L755

`ColorTransfer` is normalized to `VideoRange` and `VideoRangeType`. We should eventually avoid using `ColorTransfer` directly.

**Review by Shadowghost — APPROVED** [review]

**Pull request merged**